### ISSUE TITLE
Disk-based Storage

### ISSUE THREAD

**Issue by tsandall** [open-policy-agent/opa#4014]
Now that OPA includes a disk-based implementation of the storage.Store interface, we can work on integrating the implementation into the OPA server. This issue tracks the work that needs to get done and will need to be updated as there are probably unknowns:

- [x] identify and refactor internal code that relies on reading from transaction in the commit hook (this isn't supported by the disk-based store)
- [x] identify and refactor internal code that relies on data paths in transaction events (or improve disk-based store to record data paths for transactions)
- [x] expose configuration to enable disk-based store (minimally storage directory and partitions)
- [x] instantiate disk-based store in runtime instead of in-memory store based on configuration
- [ ] investigate optimizations to avoid loading bundles fully into memory and instead stream-to-disk
- [x] investigate whether server integration tests can be run with disk-based storage to show conformance

**Comment by floriangasc** [comment]
@tsandall 

Hi, 

If you need help for something may be i can help.  
1. `expose configuration to enable disk-based store (minimally storage directory and partitions)` :  I have also make a little code for add to cli args a disk configuration choice.
2. ` instantiate disk-based store in runtime instead of in-memory store based on configuration` : I started to write a little function for chose store type in runtime according config may be it can use at base. 

If you want i can copy past code (or make a pr) if it could be useful. It's not a big part of the problem, i think.

For 2 first point (refactoring) is may be little hard for me. For optimisation  stream-to-disk, i have take a look but it's seem's there lot of code involve and it's also little hard. But if i can help you on some task why not.

**Comment by tsandall** [comment]
Hi @floriangasc - I agree that the first two points are probably the hardest. I think we need to get those sorted out first since they're the highest risk. I'm hoping that we'll work on this in the next 2-3 months.

**Comment by floriangasc** [comment]
Hi @tsandall 

I will try to use in production mirroring open policy agent with disk store enabled. Here some feed back. I don't know if you already know, but if it could be help:

1. the disk store implementation, badger, take not negligible ram especially in case of write occur often (5-10mn). In our case, the same amount ram with and without badger is consumed. It's not for the same reason but at the end in OS point of view, it's same amount of ram. The main reason i have found is about `memtable` of badger (thanks to pprof). I have tried conservative ram configuration found on some issue, but it consumes too much ram for use as service sidecar (greater than 6 giga).
2. I have read somewhere in our opa github what properties of badger should be customizable. I think some properties should be added. I have need to use this
```o := badger.DefaultOptions(opts.Dir).WithLogger(nil)
o = o.WithIndexCacheSize(100 << 20)
o.NumMemtables = 1
o.MemTableSize = 32 << 20
o = o.WithBaseTableSize(1 << 20)
o = o.WithBlockCacheSize(1 << 20)
```


Some useful link i have found during my research on badger.
https://github.com/dgraph-io/ristretto#Config
https://discuss.dgraph.io/t/memtable-memory-usage-causes-oom/9059
https://dgraph.io/docs/badger/get-started/#in-memory-mode-diskless-mode/
https://github.com/dgraph-io/badger/blob/master/options.go
https://github.com/dgraph-io/badger/issues/1304


The disk store ram consumption issue is not coupled to number of data loaded, it's more about a minimum ram need by badger to work (according to data loaded, and data access and data updated)


Finally after study opa github issue about this kind of problem, i have choose to develop a plugin, that use opa downloader for download bundle localy. But bundle are not activated. It's just localy present in sidecar. It's activated on demand, according input, for a defined duration (like 5 or 10mn according production feedback). Now opa consume 512 mo de ram in the worst case, policy are evaluated in 10 micro second in worst case. Bundle are keep up-to-date due to long polling. And ram consumption not depend on all user, group and so on present in database, but only those connected at the specific service. So plugin is the most promising solution, I have abandoned to use disk store, because it's cause to ram consumption.

If you want any more information, let me know. Thanks for all your work, opa has very good code (easy to read and learn and update) and it's solve a lot of problems in clean way for us.

**Comment by tsandall** [comment]
@floriangasc sorry for the delayed reply--this got buried in updates over the holidays.

Is it possible to share the code you used to stress test the disk store in OPA? I didn't observe the behaviour you're describing when I tested badger [almost a year ago](https://github.com/tsandall/opa-persistence-exp) but perhaps I missed something.

**Comment by tsandall** [comment]
@srenatus just to double check... the last task on the list above (_"investigate whether server integration tests can be run with disk-based storage to show conformance"_) is complete, right?

**Comment by srenatus** [comment]
Yes, a select few have been changed to run with both inmem and disk storage; notably those checking the `PUT /v1/data` API. They had yielded a good number of issues that had been resolved via #4381.

Agreed, I'll close this one in favour of an issue focusing on bundles.

**Comment by floriangasc** [comment]
@tsandall 

Hi. Sorry for the delay, I have forgotten to respond. I have use istio mirroring capability.
1. i have put opa has side car of dedicated ingressgateway testing  (openresty/nginx)  k8s deployment.
2. i have configure opa to "listen" our production bundle server (long polling).
3. A new bundle or new version of bundle pop every 5mn and opa evaluate around 50-100 request per second.
4. the ingress gateway testing k8s deployment get all mirrored traffic (GET requests) of real ingress gateway k8s deployment.
5. It run about five days (or something like that) after we have decide to switch to another way.
